raft_group0: start leadership monitor fiber during restart

patjed41 opened the pull request:
Currently, we start group 0 leadership monitor fiber only during
a normal bootstrap. However, we should also do it when we restart
a node (either with or without upgrading it to Raft).

Fixes #15166

patjed41 commented:
v2:
- properly handle restart with upgrade
- handle restart without upgrade in `setup_group0_if_exist` instead of `setup_group0`

patjed41 commented:
@kbr-scylla ping

kbr-scylla commented on service/raft/raft_group0.cc:
```diff
@@ -402,13 +402,15 @@ future<> raft_group0::leadership_monitor_fiber() {
+                group0_log.info("losing leadership");
+            } else if (group0_server().is_leader()) {
+                group0_log.info("gaining leadership");
+            }
```
I'm not sure this is correct.
Suppose that the server was a leader at the beginning of the loop. We `wait_for_state_change`. Then the server loses leadership and quickly gains it back, before `wait_for_state_change` wakes up.

In that case we will print `losing_leadership` and won't print `gaining_leadership`, then enter another wait.

The previous version didn't have this problem.

Is printing "gaining leadership" after being chosen discovery leader really a problem? I wouldn't say. I think it's pretty consistent -- the server always prints "gaining leadership" when becoming a group 0 leader, irrespective of why it happened.

patjed41 replied:
The problem is that "gaining leadership" is not printed instantly after being chosen as the discovery leader because `leadership_monitor_fiber` is not even started. A mistimed "gaining leadership" log might be confusing. But even if it is a problem, it's much smaller than the incorrectness described by you. I'll remove the second commit.

patjed41 commented:
v3:
- removed the second commit (the one with a change that ignored a discovery leader)
- since this PR has only one commit now, changed its name and description to match the single commit

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3416/
